### ISSUE TITLE
File system restrictions in GUI action?

### ISSUE THREAD

**Issue by Beechside** [supercollider/supercollider#3306]
Clicking the button in the following results in an error but arguably shouldn't:

`Button(Window.new.front).action = {thisProcess.nowExecutingPath.dirname.postln};`

The following does work, however:

```
a = thisProcess.nowExecutingPath.dirname;
Button(Window.new.front).action = {a.postln};
```

Is this a deliberate restriction or a bug?

Supercollider 3.9dev; Mac OS X 10.13.1

**Comment by brianlheim** [comment]
cc @jamshark70 

I think this is expected behavior, because the code is being scheduled, i.e. not run interactively. `nowExecutingPath` can only be non-nil while the interpreter is running.

**Comment by Beechside** [comment]
Thank you for your input.

That was my first thought but, unless the interpreter converts all the client-side code (effectively to some sort of pcode - which I don't believe it does), anything that is event driven, rather than procedural, should still get interpreted when the event occurs.  I always assumed the interpreter was a classical line-by line interpreter and was called as required.

So, I still don't see why it can't resolve `nowExecutingPath` however...

**Comment by brianlheim** [comment]
SuperCollider code gets compiled to VM bytecodes which are then associated with the function object you've created (specifically, via an instance of `FunctionDef`). The specific values of objects referred to by those bytecodes are not resolved until later. For example:

```supercollider
(
var test = 3;
var func = {test.postln};
test = 4;
func.value; // prints '4'
)
```

In case you're interested, here are the opcodes for the function in question:

```supercollider
{thisProcess.nowExecutingPath.dirname.postln}.def.dumpByteCodes
```
gives
```
BYTECODES: (10)
  0   0F 0F    opgProcess
  2   A1 01    SendMsg 'nowExecutingPath'
  4   A1 00    SendMsg 'dirname'
  6   B0       TailCallReturnFromFunction
  7   C1 3A    SendSpecialMsg 'postln'
  9   F2       BlockReturn
```

I was not very clear before:

>` nowExecutingPath` can only be non-nil while the interpreter is running.

is not specific enough. I should have said:

> `nowExecutingPath` can only be non-nil while the interpreter is running interactively.

Code inside functions will be compiled, but not interpreted. When the event is triggered by pressing the button, the bytecodes in that compiled function object will then run on the interpreter. But, since the interpreter is not executing interactively, `nowExecutingPath` is `nil`. Compare that to what happens when the function is evaluated in an interactive setting:

```
{thisProcess.nowExecutingPath.dirname.postln}.value
```

**Comment by jamshark70** [comment]
TL;DR At the top of your code, save the executing path in a variable:

```
var thisPath = thisProcess.nowExecutingPath;

var myWindows, myViews, etc; ...

...
```

Beyond that: Brian is correct. SC is not a classical interpreter. When the bytecode VM enters a function, it does not look up the source code in any file. The source code has already been compiled into bytecodes, and a FunctionDef doesn't keep its source path. (FunctionDef has a sourceCode variable, which might make it look like an interpreter, but in fact, this is only for printing closed functions in `.asCompileString`.)

There isn't a strong reason why a FunctionDef could *not* hold its source path, except that using it for nowExecutingPath would involve surgery on the virtual machine, which we're all scared to do.

**Comment by Beechside** [comment]
Understood and closed.  Thank you.
